### PR TITLE
feat: Map Streamer

### LINKED ISSUE #108: Implement Map Stream Functionality

### PR DESCRIPTION
**Pull request by yhl25**
fixes #108

### CODE REVIEW

File: numaflow/src/mapstream.rs [hunk 1]
```diff
@@ -0,0 +1,887 @@
...
+                    "Failed to send response: {e:?}"
+                ))))
+                .await
+                .expect("Sending error on channel");
+            return;
+        }
+    }
+
+    // we should not end eof message if the map stream panicked
+    if panic_listener.is_err() {
```
Comment (vigith) [inline numaflow/src/mapstream.rs:400]:
```suggestion
    if panic_listener.is_err() {
        error!(err=?panic_listener, "there was a panic invoked");
```

File: numaflow/src/mapstream.rs [hunk 2]
```diff
@@ -0,0 +1,887 @@
...
+        Server {
+            sock_addr: DEFAULT_SOCK_ADDR.into(),
+            max_message_size: DEFAULT_MAX_MESSAGE_SIZE,
+            server_info_file: DEFAULT_SERVER_INFO_FILE.into(),
+            svc: Some(map_stream_svc),
+        }
+    }
+
+    /// Set the unix domain socket file path used by the gRPC server to listen for incoming connections.
+    /// Default value is `/var/run/numaflow/map.sock`
```
Comment (vigith) [inline numaflow/src/mapstream.rs:494]:
```suggestion
    /// Default value is `/var/run/numaflow/mapstream.sock`
```

File: numaflow/src/mapstream.rs [hunk 3]
```diff
@@ -0,0 +1,887 @@
...
+    pub fn with_socket_file(mut self, file: impl Into<PathBuf>) -> Self {
+        self.sock_addr = file.into();
+        self
+    }
+
+    /// Get the unix domain socket file path where gRPC server listens for incoming connections. Default value is `/var/run/numaflow/map.sock`
```
Comment (vigith) [inline numaflow/src/mapstream.rs:500]:
```suggestion
    /// Get the unix domain socket file path where gRPC server listens for incoming connections. Default value is `/var/run/numaflow/mapstream.sock`
```

**Review by vigith — APPROVED** [review]

File: numaflow/src/mapstream.rs [hunk 4]
```diff
@@ -0,0 +1,887 @@
...
+    ///
+    /// * `keys` - A vector of strings representing the keys.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    ///  use numaflow::map::Message;
+    /// let message = Message::new(vec![1, 2, 3]).keys(vec!["key1".to_string(), "key2".to_string()]);
+    /// ```
+    pub fn keys(mut self, keys: Vec<String>) -> Self {
```
Comment (BulkBeing) [inline numaflow/src/mapstream.rs:132]: use `with_` prefix for methods that set the member fields and return `Self`, like `with_keys()`, `with_tags()`

**Review by vigith — APPROVED** [review]

**Review by BulkBeing — APPROVED** [review]

**Pull request merged**